### ISSUE TITLE
Clarification needed for Use Debug Libraries option

### ISSUE THREAD

**Issue by levicki** [MicrosoftDocs/cpp-docs#2470]
The intended use of **Use Debug Libraries** option is totally unclear.

From what I noticed if set to yes it inserts a referrence to MSVCRTD.lib in the .obj file regardless of the project's build target (Debug or Release). This can lead to problems if you are building a library, or a project containing multiple 3rd party libraries which have this option configured differently.

It is also unclear which mechanism this option is using to change the behavior of the compiler and linker. What switch is this option equivalent to? What project file node or attribute? There are already `/MT`, `/MD`, and `/LD` compiler switches controlling which runtime libraries are used and how they are linked. This option partially overrides those, and yet it is completely undocumented.

TL;DR -- Having this option and not explaining it properly just creates unneccessary confusion.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 69b82557-4a26-70f9-c7ae-ca21c91318ec
* Version Independent ID: d93c9b26-7bde-7186-0d27-b37ade129b3c
* Content: [Advanced Property Page (Project)](https://docs.microsoft.com/en-us/cpp/build/reference/advanced-property-page?view=vs-2019)
* Content Source: [docs/build/reference/advanced-property-page.md](https://github.com/Microsoft/cpp-docs/blob/master/docs/build/reference/advanced-property-page.md)
* Product: **visual-cpp**
* Technology: **cpp-tools**
* GitHub Login: @corob-msft
* Microsoft Alias: **corob**

**Comment by corob-msft** [comment]
@levicki 
This property may be misnamed, and causing confusion as a result. It's used to set a constellation of properties in your project file for debug or release builds, one of which is the default library set to use. If you create your own build configurations, you can use this property as shorthand to set a build up for debug or release. While it's possible to link debug libraries into an otherwise release build, you can't ship an app that way. On the other hand, it's not unusual to create a debug build that has lots of release-build components linked in to save space. You could create a semi-debug configuration that specifies a retail build at the project level (that is, **Use Debug Libraries** is set to **no**), and link in the retail libraries, but build with debugging symbols for certain objects, for instance. The property itself is named `UseDebugLibraries`. It's found in the *`general_advanced.xml`* file in the MSBuild directory of your Visual Studio installation (for example, *`C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Microsoft\VC\v160\1033\general_advanced.xml`*). While some confusion is probably inevitable, I don't feel this article is the place for such detailed information, so I'm closing the issue.

**Comment by levicki** [comment]
Congratulations on closing the ticket and marking a "job well done" checkbox in your work queue without answering a single question I asked.

> It's used to **set a constellation of properties in your project file** for debug or release builds, one of which is the default library set to use.

> I don't feel this article is the **place for such detailed information**

So, what are all those properties that are getting set by this property?
What is the right place to document them if not the article which describes said property?

I know this is not right place to discuss the following, but introducing this property is a terrible design choice.

There is already a way to set a build for debug and release -- you choose either Debug or Release build and build it.

This property only allows you to create a Franken-build -- granted it was possible to do this before, but you were at least aware of which individual properties you changed in the project file. This changes a bunch of undocumented properties (including runtime libraries as you say) and both the mechanism of doing so as well as the priority over individual options already set in the project properties is unclear, hence my request for clarification.

**Comment by corob-msft** [comment]
@levicki 
I appreciate your frustration, but I'm not going to address it. Basically, this property is what makes a debug configuration a debug configuration. Most people will be satisfied to know that it's a one-step spot to set the compiler, linker, and library options you need to create a new debug or release build configuration. It's almost nonsensical to turn it off in a debug build configuration, or to turn it on in a release configuration. You'll only want to fiddle with it in configurations you create yourself. If you don't understand the project system well enough to trace its effects, you're best off leaving the setting alone in an existing configuration.

Everything the property does is documented in the .xml and .props files available in the VC project build system, found in the *`%VSINSTALLDIR%\MSBuild\Microsoft\VC\v160\`* directory of VS 2019. It's referred to at least 120 times across 39 files to set various debug options for different kinds of build targets and platforms. (I'm looking at my VS install; yours may differ, depending on what tools, targets, and platforms you have installed.) Rather than enumerate those effects in this article, it's best to go to the source. Sometimes a lack of documentation means "for experts only." This is one of those times.

**Comment by levicki** [comment]
If it is "for experts only" then why make it easy to change by non-experts? Make up your mind, please.

I am writing Windows software since Windows 95 and using software tools for just as long, and so far no software vendor ever told me "if you want to know how our product works just reverse-engineer it", much less "if you can't figure it out on your own, then you are stupid and shouldn't use it".

If something is complex but easily accessible to everyone then it deserves better explanation, not a condenscending snark.

TL;DR -- From the current docs and the property name itself it is not even clear that it changes more than just runtime libraries.

The least you could do is add a line saying "This property is a shorthand for setting build configuration as Debug or Release, for the list of affected configuration properties see whatever.xml in such and such folder".

Or you could leave it as is and keep calling paying customers idiots.

**Comment by allan-bommer** [comment]
Same question here....don't know why it can't be answered...

**Comment by trueqbit** [comment]
+1
@corob-msft If this is labeled as 'doc-enhancement' then why is this ticket closed?

**Comment by jesko42** [comment]
Looks like MS employers tend to hide unability to change things. Since 30 years they are unable to write helpful help-documents as well as integrate themselfs into the community to leverage progress.

Same here with [Billy Robert O'Neal III]: https://developercommunity.visualstudio.com/t/support-long-paths-by-default-in-vs-implementation/377910

I tell you: This is a Microsoft-Pattern!

**Comment by gh-andre** [comment]
@colin-home 

> The property itself is named `UseDebugLibraries`. It's found in the _`general_advanced.xml`_ file in the MSBuild directory of your Visual Studio installation (for example, _`C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Microsoft\VC\v160\1033\general_advanced.xml`_).

Thank you for this description and the directory location. `UseDebugLibraries` is exactly what I was looking for while building Nuget packages that have fewer configurations than the project that uses them has, which requires configuration mappings. Trying to figure out whether it's a debug or a release build from all other settings, in order to pick a debug/release configuration, proved to be a challenge because of how properties work out in MSBuild projects, but this one captures all of those properties in one place.
